atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/ScalingSweepTaskSchedulerTest.java

```diff
@@ -0,0 +1,310 @@
+        verify(sweepIteration, times(1 + 5)).call();
+    }
+
+    @Test
+    public void whenFewEntriesAreSweptTasksAreReduced() throws Exception {
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL, SUCCESS_MEDIUM);
+        scheduler.start(10);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(10 + 5 * 9)).call();
+    }
+
+    @Test
+    public void whenFewEntriesDoNotReduceToZeroTasks() throws Exception {
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL);
+        scheduler.start(1);
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(1 + 5)).call();
+    }
+
+    @Test
+    public void coolDownProtectsAgainstReducingTasks() throws Exception {
+        ScalingSweepTaskScheduler schedulerWithCoolDown = createScheduler(Duration.ofDays(1));
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL);
+
+        schedulerWithCoolDown.start(2);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(2 + 5 * 2)).call();
+    }
+
+    @Test
+    public void attemptToIncreaseNumberOfThreadsPreventsReduction() throws Exception {
+        Duration coolDown = Duration.ofSeconds(1);
+        ScalingSweepTaskScheduler schedulerWithCoolDown = createScheduler(coolDown);
+        when(sweepIteration.call()).thenReturn(
+                SUCCESS_LARGE,
+                SUCCESS_LARGE,
+                SUCCESS_SMALL);
+        schedulerWithCoolDown.start(2);
+
+        Uninterruptibles.sleepUninterruptibly(coolDown.toMillis() / 2 + 1, TimeUnit.MILLISECONDS);
+        deterministicScheduler.tick(INITIAL_DELAY, TimeUnit.MILLISECONDS);
+
+        Uninterruptibles.sleepUninterruptibly(coolDown.toMillis() / 2 + 1, TimeUnit.MILLISECONDS);
+        // these will not reduce the number of threads due to the attempt to increase in previous iterations
+        deterministicScheduler.tick(2 * DELAY, TimeUnit.MILLISECONDS);
+
+        Uninterruptibles.sleepUninterruptibly(coolDown.toMillis() / 2 + 1, TimeUnit.MILLISECONDS);
+        // the first iteration that gets executed will not reschedule
+        deterministicScheduler.tick(2 * DELAY, TimeUnit.MILLISECONDS);
+
+        verify(sweepIteration, times(2 + 2 * 2 + 2 + 1)).call();
+    }
+
+    @Test
+    public void whenUnableToAcquireShardOnLastTaskRescheduleAfterMaxPause() throws Exception {
+        ScalingSweepTaskScheduler schedulerWithRealDelay = createScheduler(new SweepDelay(1L), Duration.ZERO);
+        when(sweepIteration.call()).thenReturn(SweepIterationResults.unableToAcquireShard(), SUCCESS_MEDIUM);
+
+        schedulerWithRealDelay.start(1);
+        deterministicScheduler.tick(INITIAL_DELAY + SweepDelay.DEFAULT_MAX_PAUSE_MILLIS + 2, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(1 + 3)).call();
```

Comment:
In general for tests like this one, you might want to tick the ticker a bit less, check that it hadn't retriggered yet, and then tick it to the required value